### PR TITLE
feat(plasma-ui): add theme provider for ssg

### PR DESCRIPTION
**Pull request by RPG59**
:baby_chick: Download canary assets:

[colors--canary.267.3749489941.xml](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/colors--canary.267.3749489941.xml)  
[color_metro_ios-swift--canary.267.3749489941.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_ios-swift--canary.267.3749489941.swift)  
[color_metro_kotlin--canary.267.3749489941.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_kotlin--canary.267.3749489941.kt)  
[color_metro_react-native--canary.267.3749489941.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_react-native--canary.267.3749489941.ts)  
[color_sbermarket_ios-swift--canary.267.3749489941.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_ios-swift--canary.267.3749489941.swift)  
[color_sbermarket_kotlin--canary.267.3749489941.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_kotlin--canary.267.3749489941.kt)  
[color_sbermarket_react-native--canary.267.3749489941.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_react-native--canary.267.3749489941.ts)  
[color_sberprime_ios-swift--canary.267.3749489941.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_ios-swift--canary.267.3749489941.swift)  
[color_sberprime_kotlin--canary.267.3749489941.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_kotlin--canary.267.3749489941.kt)  
[color_sberprime_react-native--canary.267.3749489941.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_react-native--canary.267.3749489941.ts)  
[color_selgros_ios-swift--canary.267.3749489941.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_ios-swift--canary.267.3749489941.swift)  
[color_selgros_kotlin--canary.267.3749489941.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_kotlin--canary.267.3749489941.kt)  
[color_selgros_react-native--canary.267.3749489941.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_react-native--canary.267.3749489941.ts)  
[color_smbusiness_ios-swift--canary.267.3749489941.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_ios-swift--canary.267.3749489941.swift)  
[color_smbusiness_kotlin--canary.267.3749489941.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_kotlin--canary.267.3749489941.kt)  
[color_smbusiness_react-native--canary.267.3749489941.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_react-native--canary.267.3749489941.ts)  
[PlasmaTokensColor--canary.267.3749489941.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/PlasmaTokensColor--canary.267.3749489941.swift)  
[typo_mage_kotlin--canary.267.3749489941.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_kotlin--canary.267.3749489941.kt)  
[typo_mage_react-native--canary.267.3749489941.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_react-native--canary.267.3749489941.ts)  
[typo_plasma_kotlin--canary.267.3749489941.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_kotlin--canary.267.3749489941.kt)  
[typo_plasma_react-native--canary.267.3749489941.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_react-native--canary.267.3749489941.ts)  
[typo_ruler_kotlin--canary.267.3749489941.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_kotlin--canary.267.3749489941.kt)  
[typo_ruler_react-native--canary.267.3749489941.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_react-native--canary.267.3749489941.ts)  
[typo_sage_kotlin--canary.267.3749489941.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_kotlin--canary.267.3749489941.kt)  
[typo_sage_react-native--canary.267.3749489941.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_react-native--canary.267.3749489941.ts)  
[typo_sbermarket_kotlin--canary.267.3749489941.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_kotlin--canary.267.3749489941.kt)  
[typo_sbermarket_react-native--canary.267.3749489941.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_react-native--canary.267.3749489941.ts)  
[typo_soulmate_kotlin--canary.267.3749489941.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_kotlin--canary.267.3749489941.kt)  
[typo_soulmate_react-native--canary.267.3749489941.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_react-native--canary.267.3749489941.ts)

<details>
  <summary>📦 Published PR as canary version: <code>Canary Versions</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @salutejs/plasma-temple@1.119.0-canary.267.3749489941.0
  npm install @salutejs/plasma-ui@1.152.0-canary.267.3749489941.0
  # or 
  yarn add @salutejs/plasma-temple@1.119.0-canary.267.3749489941.0
  yarn add @salutejs/plasma-ui@1.152.0-canary.267.3749489941.0
  ```
</details>

### CODE REVIEW

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-267/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-267/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-267/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-267/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-267/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-267/temple-storybook/

File: packages/plasma-ui/src/components/Device/DeviceDetection.tsx [hunk 1]
```diff
@@ -83,9 +70,21 @@ export const DeviceThemeProvider = ({
...
+export const DeviceThemeProviderSSG = ({
+    theme,
+    children,
+    detectDeviceCallback = detectDevice,
+    lowPerformance = false,
+}: DeviceThemeProps) => {
+    const deviceKind = detectDeviceCallback();
+    const deviceScale = deviceScales[deviceKind] || sberPortalScale;
+
+    return <ThemeProvider theme={{ ...theme, deviceScale, lowPerformance, deviceKind }}>{children}</ThemeProvider>;
```
Comment (arsenykruglikov) [inline packages/plasma-ui/src/components/Device/DeviceDetection.tsx:89]: А разве не надо замемоизировать `{ ...theme, deviceScale, lowPerformance, deviceKind }`?

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-267/

File: packages/plasma-ui/src/components/Device/DeviceDetection.tsx [hunk 2]
```diff
@@ -83,9 +70,27 @@ export const DeviceThemeProvider = ({
...
+
+export const DeviceThemeProviderSSG = ({
+    theme,
+    children,
+    detectDeviceCallback = detectDevice,
+    lowPerformance = false,
+}: DeviceThemeProps) => {
+    const deviceKind = detectDeviceCallback();
+    const deviceScale = deviceScales[deviceKind] || sberPortalScale;
+    const memoizedThemeData = useMemo(() => ({ ...theme, deviceScale, lowPerformance, deviceKind }), [
```
Comment (akhdrv) [inline packages/plasma-ui/src/components/Device/DeviceDetection.tsx:88]: а зачем мемоизировать, если это не будет гидрироваться? похоже что мемоизировать нужно в компоненте выше

File: packages/plasma-ui/src/components/Typo/index.tsx [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+:root {
+    ${transformStyles(typo[':root'])}
+}`;
+
+const TypoSberbox = createGlobalStyle`${transformWithRoot(sberBox)}`;
+const TypoMobile = createGlobalStyle`${transformWithRoot(mobile)}`;
+const TypoPortal = createGlobalStyle`${transformWithRoot(sberPortal)}`;
+/* stylelint-enable */
+
+export const Typo = ({ deviceKind }: { deviceKind: DeviceKind }) => {
```
Comment (akhdrv) [inline packages/plasma-ui/src/components/Typo/index.tsx:18]: зачем выносить и зачем маппинг разворачивать? там бы useMemo убрать да и все

File: packages/plasma-ui/src/components/Typo/index.tsx [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import React from 'react';
```
Comment (KateKate) [inline packages/plasma-ui/src/components/Typo/index.tsx:1]: Оставь плиз по месту использования. Не надо в отдельный файл столько мало строчек кода + используется в одном месте - усложняет чтение

File: packages/plasma-ui/src/components/Device/DeviceDetection.tsx (hunk 2 above)
Comment (KateKate) [reply packages/plasma-ui/src/components/Device/DeviceDetection.tsx:88]: в теории я так понимаю это мемоизируем, чтобы [theme.context на каждый рендер нового не было](https://github.com/styled-components/styled-components/blob/bb57780f22209920f9ef61b34f1ada02bddfa878/packages/styled-components/src/models/ThemeProvider.tsx#L74) Но если правда применение только для renderToString, зачем мемоизировать.

Вань, и давай комменты напишем, что это для renderToString и пока экспериментальная фича (а может и нет, я забыла почему ты не хочешь в доку пока это добавлять)

File: packages/plasma-ui/src/components/Typo/index.tsx [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    ${transformStyles(typo[':root'])}
+}`;
+
+const TypoSberbox = createGlobalStyle`${transformWithRoot(sberBox)}`;
+const TypoMobile = createGlobalStyle`${transformWithRoot(mobile)}`;
+const TypoPortal = createGlobalStyle`${transformWithRoot(sberPortal)}`;
+/* stylelint-enable */
+
+export const Typo = ({ deviceKind }: { deviceKind: DeviceKind }) => {
+    if (deviceKind === 'mobile') {
```
Comment (KateKate) [inline packages/plasma-ui/src/components/Typo/index.tsx:19]: лучше через объект как было - хороший подход - лучше читается и строчек кода меньше

File: packages/plasma-ui/src/components/Device/DeviceDetection.tsx [hunk 6]
```diff
@@ -83,9 +70,27 @@ export const DeviceThemeProvider = ({
...
             ) : (
-                <Typo />
+                <Typo deviceKind={deviceKind} />
             )}
             {children}
         </ThemeProvider>
     );
 };
+
+export const DeviceThemeProviderSSG = ({
```
Comment (KateKate) [inline packages/plasma-ui/src/components/Device/DeviceDetection.tsx:80]: давай плиз назовем DeviceThemeProviderSSG => DeviceThemeProviderWOTypo и комменты для чего оно так. Оно не только для SSG полезно, но еще и для SSR

File: packages/plasma-ui/src/components/Typo/index.tsx [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+import { DeviceKind, transformStyles } from '../../utils';
+
+/* stylelint-disable */
+const transformWithRoot = (typo: typeof sberBox) => `
+:root {
+    ${transformStyles(typo[':root'])}
+}`;
+
+const TypoSberbox = createGlobalStyle`${transformWithRoot(sberBox)}`;
```
Comment (KateKate) [inline packages/plasma-ui/src/components/Typo/index.tsx:13]: кажется тут везде export надо, чтобы в сервисах использовать

File: packages/plasma-ui/src/components/Device/DeviceDetection.tsx [hunk 8]
```diff
@@ -72,7 +60,6 @@ export const DeviceThemeProvider = ({
 }: DeviceThemeProps) => {
     const deviceKind = detectDeviceCallback();
     const deviceScale = deviceScales[deviceKind] || sberPortalScale;
-    const Typo = React.useMemo(() => typoSizes[deviceKind], [deviceKind]);
```
Comment (KateKate) [inline packages/plasma-ui/src/components/Device/DeviceDetection.tsx:75]: кажется можно чтобы DeviceThemeProvider переиспользовал DeviceThemeProviderSSG => копи пасты меньше +логика очевиднее

File: packages/plasma-ui/src/components/Typo/index.tsx (hunk 7 above)
Comment (RPG59) [reply packages/plasma-ui/src/components/Typo/index.tsx:13]: в сервисах использовать `gloablStyle` ?

**Pull request closed without merging**